benchmark/ext_helpers/sklearn.py

```diff
@@ -0,0 +1,26 @@
+from sklearn.cluster._kmeans import _kmeans_single_lloyd, row_norms
+
+
+def kmeans(
+    X,
+    sample_weight,
+    centers_init,
+    max_iter=300,
+    verbose=False,
+    x_squared_norms=None,
+    tol=1e-4,
+    n_threads=1,
+):
+    # NB: include computation of x_squared_norms here to be fairer to other
+    # implementations that don't make use of x_squared_norms input
```

Comment:
In sklearn `x_squared_norms` is pre-computed and passed to the `_kmeans_single_lloyd` function.

The benchmark only measure the performance of `_kmeans_single_lloyd`.

Compared to the other implementations, vanilla `_kmeans_single_lloyd` is the only one that actually make use of this input `x_squared_norms`.

I thought it would be more accurate that for `_kmeans_single_lloyd` computation time of `x_squared_norms` is counted in the total compute time.

I thought of this detail when writing the benchmark code but in practice it does not significantly impact the results anyway.